File: src/backend/routes/auth_routes.py
from collections import defaultdict
from datetime import datetime, timedelta, timezone

from flask import Blueprint, g, jsonify, request

from auth import (
    DUMMY_HASH,
    create_session,
    current_user,
    destroy_session,
    hash_password,
    login_required,
    revoke_user_sessions,
    verify_password,
)
from config import config
from database import audit, execute, query_one, security_event, utc_now

auth_bp = Blueprint("auth", __name__, url_prefix="/api/auth")
developer_bp = Blueprint("developer", __name__, url_prefix="/api/developer")

PORTAL_ROLES = {
    "STUDENT": ("STUDENT",),
    "STAFF": ("STAFF",),
    "ADMIN": ("SUPER_ADMIN", "DEVELOPER"),
}

MONITORING_EVENTS = {
    "FULLSCREEN_EXIT",
    "TAB_VISIBILITY_CHANGE",
    "NAVIGATION_ATTEMPT",
    "CONTEXT_MENU_BLOCKED",
    "COPY_BLOCKED",
}

MIN_PASSWORD_LENGTH = 8

# In-memory login throttle keyed by user + IP. Acceptable for a single-site
# event; upgrade to a shared store if the app ever runs multi-process.
_FAILURES: dict = defaultdict(list)


def _login_key(user_id: str) -> str:
    return f"{user_id}|{request.remote_addr or 'local'}"


def _check_throttled(user_id: str) -> bool:
    now = datetime.now(timezone.utc)
    key = _login_key(user_id)
    window = timedelta(minutes=config.LOGIN_WINDOW_MINUTES)
    _FAILURES[key] = [t for t in _FAILURES[key] if now - t < window]
    return len(_FAILURES[key]) >= config.LOGIN_MAX_ATTEMPTS


def _record_failure(user_id: str) -> None:
    _FAILURES[_login_key(user_id)].append(datetime.now(timezone.utc))


def _record_success(user_id: str) -> None:
    _FAILURES.pop(_login_key(user_id), None)


def public_user(row):
    return {
        "id": row["id"],
        "role": row["role"],
        "name": row["name"],
        "email": row["email"],
        "active": bool(row["active"]),
    }


@auth_bp.post("/login")
def login():
    payload = request.get_json(silent=True) or {}
    user_id = (payload.get("userId") or "").strip().upper()
    password = payload.get("password") or ""
    portal = payload.get("portal") or ""

    if not user_id or not password:
        return jsonify({"message": "Enter your ID and password to continue."}), 400
    if _check_throttled(user_id):
        security_event(
            "LOGIN_THROTTLED", user_id, "SYSTEM",
            f"Login blocked {config.LOGIN_MAX_ATTEMPTS}+ failures in "
            f"{config.LOGIN_WINDOW_MINUTES} minutes",
        )
        return jsonify({"message": "Too many attempts. Try again later."}), 429

    user = query_one("SELECT * FROM users WHERE UPPER(id) = ?", (user_id,))
    # Always run the hash comparison so response time does not reveal whether
    # an ID is registered (defeats account enumeration).
    valid = user is not None and verify_password(password, user["password_hash"])
    if user is None:
        verify_password(password, DUMMY_HASH)

    if user is None or not valid:
        _record_failure(user_id)
        security_event("LOGIN_FAILURE", user_id, "SYSTEM", f"Failed login on {portal} portal")
        return jsonify({"message": "Invalid credentials. Please check your ID and password."}), 401
    if not user["active"]:
        _record_failure(user_id)
        return jsonify({"message": "This account has been deactivated."}), 403
    if portal and user["role"] not in PORTAL_ROLES.get(portal, ()):
        _record_failure(user_id)
        security_event("LOGIN_FAILURE", user["id"], user["role"], f"Wrong portal ({portal})")
        return jsonify({"message": "This account cannot sign in from this portal."}), 403

    existing = query_one("SELECT COUNT(*) AS n FROM sessions WHERE user_id = ?", (user["id"],))
    if existing and existing["n"] > 0:
        security_event(
            "LOGIN_BLOCKED",
            user["id"],
            user["role"],
            f"Already logged in — blocked sign-in from {portal} portal",
        )
        return jsonify({"message": "You are already signed in on another device. Please log out there first."}), 409

    _record_success(user_id)
    token = create_session(user["id"], user["role"])
    security_event("LOGIN_SUCCESS", user["id"], user["role"], f"Signed in via {portal}")
    security_event("SESSION_STARTED", user["id"], user["role"], "Session started")
    return jsonify({"token": token, "user": public_user(user)})


@auth_bp.post("/logout")
def logout():
    user = current_user()
    if user is not None:
        destroy_session(g.session_token)
        security_event("SESSION_ENDED", user["id"], user["role"], "Signed out")
    return jsonify({"ok": True})


@auth_bp.get("/me")
@login_required
def me():
    return jsonify({"user": public_user(g.user)})


@auth_bp.post("/security-event")
@login_required
def report_event():
    payload = request.get_json(silent=True) or {}
    event_type = payload.get("type")
    if event_type not in MONITORING_EVENTS:
        return jsonify({"message": "Unknown monitoring signal."}), 400
    detail = (payload.get("detail") or "")[:500]
    attempt_id = payload.get("attemptId")
    if attempt_id:
        # A student may only attach monitoring signals to their own attempt.
        attempt = query_one(
            "SELECT 1 FROM attempts WHERE id = ? AND student_id = ?",
            (attempt_id, g.user["id"]),
        )
        if g.user["role"] == "STUDENT" and attempt is None:
            return jsonify({"message": "Unknown attempt."}), 400
    security_event(
        event_type,
        g.user["id"],
        g.user["role"],
        detail,
        attempt_id=attempt_id,
    )
    return jsonify({"ok": True})


def _validate_password(password: str):
    if len(password) < MIN_PASSWORD_LENGTH:
        return {
            "message": f"Password must be at least {MIN_PASSWORD_LENGTH} characters long."
        }
    return None


@developer_bp.post("/init-super-admin")
@login_required
def init_super_admin():
    """Developer bootstrap. Not exposed as a normal application feature."""
    if g.user["role"] != "DEVELOPER":
        return jsonify({"message": "Only the developer account can initialise a Super Admin."}), 403
    payload = request.get_json(silent=True) or {}
    user_id = (payload.get("id") or "").strip().upper()
    name = payload.get("name")
    password = payload.get("password") or ""
    if not user_id or not name or not password:
        return jsonify({"message": "ID, name and password are required."}), 400
    error = _validate_password(password)
    if error:
        return jsonify(error), 400
    if query_one("SELECT 1 FROM users WHERE id = ?", (user_id,)):
        return jsonify({"message": "That ID already exists."}), 409
    execute(
        "INSERT INTO users (id, role, name, email, password_hash, active, created_at)"
        " VALUES (?, 'SUPER_ADMIN', ?, ?, ?, 1, ?)",
        (user_id, name, payload.get("email", ""), hash_password(password), utc_now()),
    )
    audit(g.user["id"], g.user["role"], "Initialised Super Admin", user_id)
    return jsonify({"ok": True})

File: src/services/server/authRoutes.ts
import { audit, db, hashPassword, now, security, uid, verifyPassword } from '../db';
import type { Role } from '../../types';
import { HttpError, requireUser, type Handler } from './types';

const portalRoles: Record<string, Role[]> = {
  STUDENT: ['STUDENT'],
  STAFF: ['STAFF'],
  ADMIN: ['SUPER_ADMIN', 'DEVELOPER']
};

function publicUser(u: {id: string;role: Role;name: string;email: string;active: boolean;}) {
  return { id: u.id, role: u.role, name: u.name, email: u.email, active: u.active };
}

export const authRoutes: Record<string, Handler> = {
  'POST /api/auth/login': (ctx) => {
    const { userId, password, portal } = ctx.body ?? {};
    if (!userId || !password) {
      throw new HttpError(400, 'Enter your ID and password to continue.');
    }
    const user = db.users.find((u) => u.id.toUpperCase() === String(userId).toUpperCase());
    if (!user || !verifyPassword(String(password), user.passwordHash)) {
      security('LOGIN_FAILURE', String(userId), 'SYSTEM', `Failed login attempt on ${portal} portal`);
      throw new HttpError(401, 'Invalid credentials. Please check your ID and password.');
    }
    if (!user.active) {
      security('LOGIN_FAILURE', user.id, user.role, 'Login attempt on a deactivated account');
      throw new HttpError(403, 'This account has been deactivated. Contact examination staff.');
    }
    const allowed = portalRoles[portal as string] ?? [];
    if (portal && !allowed.includes(user.role)) {
      security('LOGIN_FAILURE', user.id, user.role, `Wrong portal used (${portal})`);
      throw new HttpError(403, 'This account cannot sign in from this portal.');
    }

    const existing = db.sessions.filter((s) => s.userId === user.id);
    if (existing.length > 0) {
      security(
        'MULTIPLE_SESSION_DETECTED',
        user.id,
        user.role,
        'A second session was opened while another was already active'
      );
    }
    const token = uid('SESS');
    db.sessions.push({ token, userId: user.id, role: user.role, createdAt: now() });
    security('LOGIN_SUCCESS', user.id, user.role, `Signed in via ${portal ?? 'portal'}`);
    security('SESSION_STARTED', user.id, user.role, `Session ${token} started`);
    return { token, user: publicUser(user) };
  },

  'POST /api/auth/logout': (ctx) => {
    if (ctx.token) {
      const idx = db.sessions.findIndex((s) => s.token === ctx.token);
      if (idx >= 0) db.sessions.splice(idx, 1);
    }
    if (ctx.user) security('SESSION_ENDED', ctx.user.id, ctx.user.role, 'Signed out');
    return { ok: true };
  },

  'GET /api/auth/me': (ctx) => ({ user: publicUser(requireUser(ctx)) }),

  'POST /api/auth/change-password': (ctx) => {
    const user = requireUser(ctx);
    const { currentPassword, newPassword } = ctx.body ?? {};
    const current = currentPassword || '';
    const next = newPassword || '';
    if (next.length < 8) throw new HttpError(400, 'Password must be at least 8 characters long.');
    const record = db.users.find((u) => u.id === user.id);
    if (!record) throw new HttpError(404, 'Account not found.');
    if (!verifyPassword(current, record.passwordHash)) {
      security('PASSWORD_CHANGE_FAILED', user.id, user.role, 'Wrong current password');
      throw new HttpError(400, 'Your current password is incorrect.');
    }
    if (verifyPassword(next, record.passwordHash)) {
      throw new HttpError(400, 'New password must be different.');
    }
    record.passwordHash = hashPassword(next);
    security('PASSWORD_CHANGED', user.id, user.role, 'Password changed');
    audit(user.id, user.role, 'Changed password', user.id, '');
    return { ok: true };
  },

  'POST /api/auth/security-event': (ctx) => {
    const user = requireUser(ctx);
    const { type, detail, attemptId } = ctx.body ?? {};
    const allowed = [
    'FULLSCREEN_EXIT',
    'TAB_VISIBILITY_CHANGE',
    'NAVIGATION_ATTEMPT',
    'CONTEXT_MENU_BLOCKED',
    'COPY_BLOCKED'];

    if (!allowed.includes(type)) throw new HttpError(400, 'Unknown monitoring signal.');
    security(type, user.id, user.role, String(detail ?? ''), { attemptId });
    return { ok: true };
  },

  /** Developer bootstrap: initialise the first Super Admin. */
  'POST /api/developer/init-super-admin': (ctx) => {
    const user = requireUser(ctx);
    if (user.role !== 'DEVELOPER') {
      throw new HttpError(403, 'Only the developer account can initialise a Super Admin.');
    }
    const { id, name, email, password } = ctx.body ?? {};
    if (!id || !name || !password) throw new HttpError(400, 'ID, name and password are required.');
    if (db.users.some((u) => u.id === id)) throw new HttpError(409, 'That ID already exists.');
    db.users.push({
      id: String(id).toUpperCase(),
      role: 'SUPER_ADMIN',
      name: String(name),
      email: String(email ?? ''),
      active: true,
      createdAt: now(),
      passwordHash: hashPassword(String(password))
    });
    audit(user.id, user.role, 'Initialised Super Admin', String(id), String(email ?? ''));
    return { ok: true };
  }
};